spec/models/dialog/terraform_template_service_dialog_spec.rb

```diff
@@ -0,0 +1,206 @@
+      let(:extra_vars) do
+        {
+          'some_extra_var'  => {:default => 'blah'},
+          'other_extra_var' => {:default => {'name' => 'some_value'}},
+          'array_extra_var' => {:default => [{'name' => 'some_value'}]}
+        }
+      end
+
+      let(:terraform_template) { terraform_template_with_no_input_vars }
+
+      it "creates a dialog with extra variables" do
+        dialog = subject.create_dialog(dialog_label, terraform_template, extra_vars)
+        expect(dialog).to have_attributes(:label => dialog_label, :buttons => "submit,cancel")
+
+        group = assert_variables_tab(dialog)
+        assert_extra_variables_group(group)
+      end
+    end
+
+    shared_examples_for "create_dialog with place-holder variable argument" do
+      it "when no terraform template input vars and empty extra vars" do
+        dialog = described_class.create_dialog(dialog_label, terraform_template, extra_vars)
+        expect(dialog).to have_attributes(:label => dialog_label, :buttons => "submit,cancel")
+
+        group = assert_variables_tab(dialog)
+        assert_default_variables_group(group, dialog_label)
+      end
+    end
+
+    context "when empty terraform template input vars & empty extra vars" do
+      let(:dialog_label) { "mydialog2" }
+
+      let(:terraform_template) { terraform_template_with_no_input_vars }
+
+      let(:extra_vars) do
+        {}
+      end
+
+      it_behaves_like "create_dialog with place-holder variable argument"
+    end
+
+    context "when nil terraform template & nil extra vars" do
+      let(:dialog_label) { "mydialog3" }
+
+      let(:terraform_template) { nil }
+
+      let(:extra_vars) { nil }
+
+      it_behaves_like "create_dialog with place-holder variable argument"
+    end
+
+    context "with terraform template input vars and with extra vars" do
+      let(:dialog_label) { "mydialog4" }
+
+      let(:extra_vars) do
+        {
+          'some_extra_var'  => {:default => 'blah'},
+          'other_extra_var' => {:default => {'name' => 'some_value'}},
+          'array_extra_var' => {:default => [{'name' => 'some_value'}]}
+        }
+      end
+
+      let(:input_vars) do
+        require 'json'
+        payload = JSON.parse(payload_with_three_input_vars)
+        payload['input_vars']
+      end
+
+      it "creates multiple dialogs" do
+        dialog = subject.create_dialog(dialog_label, terraform_template_with_input_vars, extra_vars)
+        expect(dialog).to have_attributes(:label => dialog_label, :buttons => "submit,cancel")
+ \
+        group1 = assert_terraform_template_variables_tab(dialog, :group_size => 2)
+        assert_terraform_variables_group(group1, input_vars)
+
+        group2 = assert_variables_tab(dialog, :group_size => 2)
+        assert_extra_variables_group(group2)
+      end
+    end
+  end
+
+  def assert_variables_tab(dialog, group_size: 1)
+    tabs = dialog.dialog_tabs
+    expect(tabs.size).to eq(1)
+
+    tab0 = tabs[0]
+    assert_tab_attributes(tab0)
```

Comment:
Minor formatting concern
```suggestion
    assert_tab_attributes(tabs.first)
```